### DISCUSSION TITLE
Event handler to check Vat ID based on VIES VAT number validation (EU)

### DISCUSSION THREAD

**Discussion by apcloic** [YetiForceCompany/YetiForceCRM#15568] [Feature improvement]
Hi,

I've created a custom pre save event handler to check if account vat id exists.
Maybe it could be a useful feature along with the duplicate vat id event handler ?

To use it, you have to add it to vtiger_eventhandlers table and put the content either in /modules/Accounts/handlers or in /custom/modules/Accounts/handlers if $LOAD_CUSTOM_FILES is set to true in /config/Performance.php

It's based on VIES and the following library, you can install it using composer or manually :
https://ec.europa.eu/taxation_customs/vies/technicalInformation.html
https://github.com/pH-7/eu-vat-validator

````
<?php
/**
 * Check EU vat id handler.
 *
 * @package Handler
 *
 * @copyright YetiForce Sp. z o.o
 * @license YetiForce Public License 4.0 (licenses/LicenseEN.txt or yetiforce.com)
 * Library : https://github.com/pH-7/eu-vat-validator
 */
/**
 * Accounts_CheckVatId_Handler class.
 */

// Optional path to Eu Vat Number Validator library if you choose to install without Composer
// require 'lib/autoloader.php'; 

use PH7\Eu\Vat\Validator;
use PH7\Eu\Vat\Provider\Europa;

class Accounts_CheckVatId_Handler
{
	/**
	 * EditViewPreSave handler function.
	 *
	 * @param App\EventHandler $eventHandler
	 */
	public function editViewPreSave(App\EventHandler $eventHandler)
	{
		$recordModel = $eventHandler->getRecordModel();
		$response = ['result' => true];
		$fieldModel = $recordModel->getModule()->getFieldByName('vat_id');

		if ($fieldModel->isViewable() && ($vat = $recordModel->get('vat_id'))) {

			$oVatValidator = new Validator(new Europa, substr($vat, 2), substr($vat, 0, 2));
			if (!$oVatValidator->check()) {
				$response = [
					'result' => false,
					'message' => App\Language::translate('LBL_VAT_DOES_NOT_EXISTS', $recordModel->getModuleName())
				];
			}
		}
		return $response;
	}
}

````

[DuplicateVatId.zip](https://github.com/YetiForceCompany/YetiForceCRM/files/7746376/DuplicateVatId.zip)

Regards,

**Comment by apcloic** [comment]
I've just found a better library.
https://github.com/ibericode/vat-bundle

Here's my basic event handler based on this library.
/custom/modules/Accounts/handlers/CheckVatId.php

````
<?php
/**
 * Check EU vat id handler.
 *
 * @package Handler
 *
 * @copyright YetiForce Sp. z o.o
 * @license YetiForce Public License 4.0 (licenses/LicenseEN.txt or yetiforce.com)
 * Library : https://github.com/pH-7/eu-vat-validator
 */
/**
 * Accounts_CheckVatId_Handler class.
 */


use Ibericode\Vat\Validator;


class Accounts_CheckVatId_Handler
{
	/**
	 * EditViewPreSave handler function.
	 *
	 * @param App\EventHandler $eventHandler
	 */
	public function editViewPreSave(App\EventHandler $eventHandler)
	{

		$recordModel = $eventHandler->getRecordModel();
		$response = ['result' => true];
		$fieldModel = $recordModel->getModule()->getFieldByName('vat_id');

		if ($fieldModel->isViewable() && ($vat = $recordModel->get('vat_id'))) {

			$validator = new Ibericode\Vat\Validator();

			if (!$validator->validateVatNumber($vat)) {
				$response = [
					'result' => false,
					'message' => App\Language::translate('LBL_VAT_DOES_NOT_EXISTS', $recordModel->getModuleName())
				];
			}
		}
		return $response;
	}
}

````

**Comment by mariuszkrzaczkowski** [comment]
You don't need a library to check this.
The system already has VIES checking, and there are functions for that
https://github.com/YetiForceCompany/YetiForceCRM/blob/9494effa184bc3e80bf1d1dc317d1659b19ff015/app/RecordCollectors/Vies.php#L115-L140
![image](https://user-images.githubusercontent.com/10198654/146833075-de416dd2-4cce-4f6a-b663-3d7c850c81b7.png)

**Reply by mariuszkrzaczkowski** [reply]
Available in modules: https://github.com/YetiForceCompany/YetiForceCRM/blob/9494effa184bc3e80bf1d1dc317d1659b19ff015/app/RecordCollectors/Vies.php#L21

https://gitdeveloper.yetiforce.com/index.php?module=Accounts&view=List&mid=51&parent=47

**Reply by apcloic** [reply]
Thank you Marius, I don't know why I didn't see the Vies module 😋
